### DISCUSSION TITLE
cannot find macro rocket_uri_macro_target_url

### DISCUSSION THREAD

**Discussion by thexeromin** [rwf2/Rocket#2489] [Questions]
I am trying to redirect the response but when I assigning the redirect url from mongodb response I am getting cannot find macro `rocket_uri_macro_target_url`.
Here is my code
```rust
#[get("/locator/<path>")]
pub fn get_locator(db: &State<MongoRepo>, path: String) -> Result<Redirect, Status> {
    let id = path;

    if id.is_empty() {
        return Err(Status::BadRequest);
    };

    let locator_details = db.get_locator(&id);

    match locator_details {
        Ok(locator) => {
            let target_url = locator.url;
            Ok(Redirect::to(uri!(target_url)))
        },
        Err(_) => Err(Status::InternalServerError),
    }
}
```

Here is the error message
```bash
error: cannot find macro `rocket_uri_macro_target_url` in this scope
  --> src/api/locator_api.rs:36:34
   |
36 |             Ok(Redirect::to(uri!(target_url)))
   |                                  ^^^^^^^^^^

error: could not compile `url-minifier` due to previous error
```

**Comment by robertwayne** [comment]
As far as I understand the `uri!` macro, you can't really use it in this way - someone please correct me if I am wrong, though!

It requires you to pass either a string literal or a named route, where the route is your actual route fn. It is a proc macro, and has to be able to generate the code at compile time. This would be impossible here, because your given path can only be known at runtime.

One possible solution off the top of my head is to match against your routes manually.

```rs
#[get("/redirect/<path>")]
pub fn redirect(path: String) -> Redirect {
    let route = match path.as_str() {
        "hello" => uri!(hello()),
        _ => uri!("/"),
    };

    Redirect::to(route)
}

#[get("/hello")]
async fn hello() -> String {
    "Hello, world!".to_string()
}
```

Alternatively, you could side-step it by using `Redirect::to(format!("/{path}"))` - note the leading slash. That said, `Redirect` states to always use `uri!` when building against named routes, so I can't comment on what the theoretical consequences would be here.
